sql/logic_test.go

```diff
@@ -0,0 +1,318 @@
+
+	"github.com/cockroachdb/cockroach/server"
+	"github.com/cockroachdb/cockroach/testutils"
+	"github.com/cockroachdb/cockroach/util/leaktest"
+)
+
+var (
+	resultsRE = regexp.MustCompile(`^(\d+)\s+values?\s+hashing\s+to\s+([0-9A-Fa-f]+)$`)
+	errorRE   = regexp.MustCompile(`^(?:statement|query)\s+error\s+(.*)$`)
+)
+
+type lineScanner struct {
+	*bufio.Scanner
+	line int
+}
+
+func newLineScanner(r io.Reader) *lineScanner {
+	return &lineScanner{
+		Scanner: bufio.NewScanner(r),
+		line:    0,
+	}
+}
+
+func (l *lineScanner) Scan() bool {
+	ok := l.Scanner.Scan()
+	if ok {
+		l.line++
+	}
+	return ok
+}
+
+type logicStatement struct {
+	pos       string
+	sql       string
+	expectErr string
+}
+
+type logicQuery struct {
+	pos             string
+	sql             string
+	expectErr       string
+	expectedValues  int
+	expectedHash    string
+	expectedResults []string
+}
+
+type logicValue string
+
+func (v *logicValue) Scan(value interface{}) error {
+	switch t := value.(type) {
+	case nil:
+		*v = "NULL"
+	case bool:
+		*v = logicValue(strconv.FormatBool(t))
+	case int64:
+		*v = logicValue(strconv.FormatInt(t, 10))
+	case float64:
+		*v = logicValue(strconv.FormatFloat(t, 'g', -1, 64))
+	case []byte:
+		*v = logicValue(t)
+	case string:
+		*v = logicValue(t)
+	case time.Time:
+		*v = logicValue(t.String())
+	default:
+		return fmt.Errorf("unexpected type: %T", value)
+	}
+	return nil
+}
+
+// logicTest executes the test cases specified in a file. The file format is
+// taken from the sqllogictest tool
+// (http://www.sqlite.org/sqllogictest/doc/trunk/about.wiki) with various
+// extensions to allow specifying errors and additional options. See
+// https://github.com/gregrahn/sqllogictest/ for a github mirror of the
+// sqllogictest source.
+//
+// TODO(pmattis): We currently cannot run the tests from sqllogictest due to
+// insufficient SQL coverage (e.g. lack of subqueries and aggregation
+// functions). We should work towards fixing that.
+type logicTest struct {
+	*testing.T
+}
+
+func (t logicTest) run(path string) {
+	file, err := os.Open(path)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer file.Close()
+	base := filepath.Base(path)
+
+	// TODO(pmattis): Add a flag to make it easy to run the tests against a local
+	// MySQL or Postgres instance.
+	srv := server.StartTestServer(nil)
+	db, err := sql.Open("cockroach", "https://root@"+srv.ServingAddr()+"?certs=test_certs")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer db.Close()
+	defer srv.Stop()
+
+	s := newLineScanner(file)
+	for s.Scan() {
+		fields := strings.Fields(s.Text())
+		if len(fields) == 0 {
+			continue
+		}
+		if strings.HasPrefix(fields[0], "#") {
+			// Skip comment lines.
+			continue
+		}
+		switch fields[0] {
+		case "statement":
+			stmt := logicStatement{pos: fmt.Sprintf("%s:%d", base, s.line)}
+			// Parse "query error <regexp>"
+			if m := errorRE.FindStringSubmatch(s.Text()); len(m) == 2 {
+				stmt.expectErr = m[1]
+			}
+			var buf bytes.Buffer
+			for s.Scan() {
```

Comment:
in the error case there is never more than one line following the "query error" line; can we just specifically read one line instead of scanning to the next empty line?